KIPs/kip-37.md

```diff
@@ -0,0 +1,746 @@
+
+**_Scenario#7 :_** The receiver implements the `KIP37TokenReceiver` interface and is the recipient of more than one balance change (e.g. `safeBatchTransferFrom` called).
+
+- All balance transfers that are referenced in a call to an `KIP37TokenReceiver` hook MUST be updated before the `KIP37TokenReceiver` hook is called on the recipient contract.
+- All transfer events MUST have been emitted to reflect current balance changes before an `KIP37TokenReceiver` hook is called on the recipient contract.
+- `onKIP37Received` or `onKIP37BatchReceived` MUST be called on the recipient as many times as necessary such that every balance change for the recipient in the scenario is accounted for.
+  - The return magic value for every hook call MUST be checked and acted upon as per “onKIP37Received rules” and “onKIP37BatchReceived rules”.
+- The `onKIP37BatchReceived` hook SHOULD be called on the recipient contract and its rules followed.
+  - See “onKIP37BatchReceived rules” for further rules that MUST be followed.
+- The `onKIP37Received` hook MAY be called on the recipient contract and its rules followed.
+  - See “onKIP37Received rules” for further rules that MUST be followed.
+
+**_Scenario#8 :_** You are the creator of a contract that implements the `KIP37TokenReceiver` interface and you forward the token(s) onto another address in one or both of `onKIP37Received` and `onKIP37BatchReceived`.
+
+- Forwarding should be considered acceptance and then initiating a new `safeTransferFrom` or `safeBatchTransferFrom` in a new context.
+  - The prescribed keccak256 acceptance value magic for the receiver hook being called MUST be returned after forwarding is successful.
+- The `_data` argument MAY be re-purposed for the new context.
+- If forwarding fails the transaction MAY be reverted.
+  - If the contract logic wishes to keep the ownership of the token(s) itself in this case it MAY do so.
+
+**_Scenario#9 :_** You are transferring tokens via a non-standard API call i.e. an implementation specific API and NOT `safeTransferFrom` or `safeBatchTransferFrom`.
+
+- In this scenario all balance updates and events output rules are the same as if a standard transfer function had been called.
+  - i.e. an external viewer MUST still be able to query the balance via a standard function and it MUST be identical to the balance as determined by `TransferSingle` and `TransferBatch` events alone.
+- If the receiver is a contract the `KIP37TokenReceiver` hooks still need to be called on it and the return values respected the same as if a standard transfer function had been called.
+  - However while the `safeTransferFrom` or `safeBatchTransferFrom` functions MUST revert if a receiving contract does not implement the KIP37TokenReceiver interface, a non-standard function MAY proceed with the transfer.
+  - See “Implementation specific transfer API rules”.
+
+#### Rules
+
+**_safeTransferFrom rules:_**
```

Comment:
What is the difference between this and the function description in the contract code?

Reply:
There's nothing else. But, I would like to keep both sides in terms of information.